File: ios/GrMob/App/HeadingSensor.swift
import CoreLocation
import Foundation

/// The iOS half of grmob's compass (core/heading.go).
///
///     core.StartHeading   ──▶ "sensor" {kind:"heading", command:"start"} ──▶ here
///     core.CurrentHeading ◀── "heading" host event ◀── here
///
/// # Why CoreLocation and not CoreMotion
///
/// The compass lives in `CLLocationManager`, not in the motion framework, and
/// it is the one place on iOS that does the magnetometer fusion and the
/// declination lookup for you. `CMMotionManager` can hand over a raw magnetic
/// field vector, but turning that into a bearing means reimplementing both,
/// and the declination half needs a world magnetic model this app has no
/// business shipping.
///
/// # Magnetic is free, true is not
///
/// `startUpdatingHeading` needs no authorization at all — the magnetic bearing
/// is not private information. `trueHeading` is: geographic north requires
/// knowing where on the planet you are, so CoreLocation reports it only when
/// location authorization has been granted, and otherwise sets it negative.
///
/// So this host asks for nothing and reports `magnetic` always, adding `true`
/// only when the value is valid. An app that wants true north asks for
/// location authorization by its own route; nothing here prompts, because a
/// permission dialog the user did not expect is a worse failure than a
/// bearing that is a few degrees off a map.
final class HeadingSensor: NSObject, CLLocationManagerDelegate {
    static let shared = HeadingSensor()

    /// The host→app channel, set by `SystemEvents.attach`.
    var report: ((String, String) -> Void)?

    private let manager = CLLocationManager()
    private var running = false
    private var lastSentAt: TimeInterval = 0

    /// A start was killed by something outside this app and can be resumed
    /// when that something changes back. Cleared by a start that gets through
    /// and by `stop`.
    ///
    /// # Which switch, given that the compass asks for nothing
    ///
    /// The magnetic bearing needs no authorization (see the type comment), so
    /// there is no per-app permission for a user to revoke — but the compass
    /// is still `CLLocationManager`'s, and Location Services switched off
    /// device-wide stops the manager for every one of its jobs. `CLError`
    /// spells that refusal `.denied`, the same code an app-level denial gets,
    /// and here it can only mean the global switch: nothing else was ever
    /// asked for.
    ///
    /// That is also why this file needs no equivalent of LocationSensor's
    /// authorization callback. There is no authorization to change, so
    /// `locationManagerDidChangeAuthorization` has nothing to say about a
    /// compass, and the foreground is not a substitute for a callback here —
    /// it is the only signal there has ever been.
    ///
    /// # Not measured, unlike the location arm
    ///
    /// The equivalent in LocationSensor was verified by driving the Settings
    /// switch on a simulator and reading lesson 4.12's panel. That experiment
    /// cannot be repeated for this file: `CLLocationManager.headingAvailable()`
    /// is false on the simulator, so `start` never gets past its own guard and
    /// there is no session for the switch to kill. What is written here is the
    /// pattern one file over, applied to a manager that is refused by the same
    /// switch for the same reason — and it is stated plainly because the arm
    /// costs one Bool and a dead compass costs the lesson that draws it.
    private var armed = false

    /// Throttle floor, matching the other two hosts: CoreLocation delivers
    /// heading updates far faster than a dial needs, and every event that
    /// reaches Go costs a full render pass.
    private let minInterval: TimeInterval = 0.066

    private override init() {
        super.init()
        manager.delegate = self
        // Degrees of change before CoreLocation bothers to tell us. This is
        // the platform's own filter and is cheaper than ours: below it, no
        // event is generated at all. Half a degree matches the notification
        // threshold Go applies on the other side (headingNotifyEpsilon), so
        // the two do not fight.
        manager.headingFilter = 0.5
        // The device's own top edge is the direction the compass reports,
        // which is what a Compass widget drawn upright on the screen means.
        manager.headingOrientation = .portrait
    }

    /// Dispatches one "sensor" system event. Unknown kinds and commands are dropped.
    func handle(_ data: [String: Any]) {
        guard (data["kind"] as? String) == "heading" else { return }
        switch data["command"] as? String {
        case "start": start()
        case "stop": stop()
        default: break
        }
    }

    private func start() {
        guard !running else { return }
        guard CLLocationManager.headingAvailable() else {
            // An iPad without a magnetometer, or the simulator, which reports
            // false here. One event, and Go stops waiting for a reading that
            // is never coming.
            //
            // Deliberately NOT armed: a device without a magnetometer will not
            // grow one, and arming it would retry on every foreground for the
            // life of the process to re-send a message Go already has.
            send(["available": false, "error": "no compass on this device"])
            return
        }
        beginUpdates()
    }

    /// Turn the compass on. The one place that does, so `start` and
    /// `retryIfArmed` cannot drift on how.
    private func beginUpdates() {
        running = true
        armed = false
        lastSentAt = 0
        // Torn down before it is built up, for the reason measured one file
        // over (LocationSensor.beginUpdates): a start on a manager that is
        // already in a delivery session is a no-op on top of that session's
        // state, so a manager that has just been refused stays refused and
        // delivers nothing. Harmless on the path where nothing was running —
        // stopping a manager that is not updating is defined to do nothing —
        // and the whole point on the retry path, which is the only path that
        // can reach here with a session the platform has already killed.
        manager.stopUpdatingHeading()
        manager.startUpdatingHeading()
        // No `acquiring` event, and that asymmetry with LocationSensor is the
        // sensor's rather than an omission: a cold GPS fix is tens of seconds
        // away, so Go has to be told the refusal is withdrawn before the first
        // reading can say so, whereas a magnetic bearing arrives in a frame or
        // two and `Available` defaults to true on it (see core.ReceiveHeading,
        // which clears Error on any available reading).
    }

    private func stop() {
        // Cleared even when nothing is running: a screen that unmounts while
        // the compass is refused must not leave it armed, or a later
        // foreground would start a magnetometer with no consumer. The same
        // argument as LocationSensor.stop, and core.StopHeading reaches here
        // on the hook's close path either way.
        let wasOn = running || armed
        running = false
        armed = false
        guard wasOn else { return }
        manager.stopUpdatingHeading()
    }

    /// Retries a start that Location Services being switched off device-wide
    /// killed. Called when the app comes back to the foreground.
    ///
    /// The foreground is the right signal for the same reason it is in
    /// LocationSensor: that switch is several taps inside Settings, so a user
    /// who changes it has necessarily left this app and come back, and there
    /// is no path to it that does not pass through here. It is self-limiting
    /// in the same way too — `armed` is only ever true after a running compass
    /// was refused, so an app that never drew one does nothing on every
    /// foreground, and one that is running does nothing either.
    func retryIfArmed() {
        guard armed, !running else { return }
        // Re-asked rather than assumed. The guard in `start` reports a device
        // with no magnetometer as permanently unavailable, but this arm is
        // only reachable on a device that HAD one and was refused, so the
        // answer here is about the hardware still being there — cheap, and the
        // alternative is starting a manager that cannot deliver.
        guard CLLocationManager.headingAvailable() else { return }
        beginUpdates()
    }

    // MARK: - CLLocationManagerDelegate

    func locationManager(_ manager: CLLocationManager, didUpdateHeading heading: CLHeading) {
        guard running else { return }

        // A negative magneticHeading means the reading is invalid — the
        // magnetometer is being interfered with, or has not settled. Dropped
        // rather than forwarded: Go would fold it into a plausible-looking
        // bearing, and a compass pointing confidently in the wrong direction
        // is worse than one that has not updated.
        guard heading.magneticHeading >= 0 else { return }

        let now = Date.timeIntervalSinceReferenceDate
        guard now - lastSentAt >= minInterval else { return }
        lastSentAt = now

        var payload: [String: Any] = [
            "magnetic": heading.magneticHeading,
            "ts": Int(now * 1000),
        ]
        // Both of these are negative-when-unknown in CoreLocation's API, and
        // both are simply omitted rather than forwarded as a negative: Go's
        // contract is that a missing key means "not known", which is exactly
        // what a negative means here.
        if heading.trueHeading >= 0 {
            payload["true"] = heading.trueHeading
        }
        if heading.headingAccuracy >= 0 {
            payload["accuracy"] = heading.headingAccuracy
        }
        send(payload)
    }

    func locationManagerShouldDisplayHeadingCalibration(_ manager: CLLocationManager) -> Bool {
        // Let the system put up its own figure-eight calibration sheet when it
        // decides the magnetometer needs it. Returning false here is the
        // choice an app makes when a modal would interrupt something more
        // important; a compass on screen is the one case where the interruption
        // is the point, and the sheet dismisses itself once calibrated.
        return true
    }

    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        // CLError.headingFailure is transient — strong local interference —
        // and CoreLocation recovers on its own, so it is not reported as
        // unavailable. Anything else means the manager has stopped, and Go
        // needs to know it is not getting readings.
        if (error as? CLError)?.code == .headingFailure { return }
        running = false

        // `.denied` is the one failure with a way back, so it is the one that
        // stays armed — the shape LocationSensor's equivalent arrived at after
        // a simulator run found a sensor that never asked again.
        //
        // The message is written here rather than taken from the error, and
        // for the reason measured one file over: `error.localizedDescription`
        // for this code is NSError's untranslated fallback, "The operation
        // couldn't be completed. (kCLErrorDomain error 1.)", and it was
        // reaching the screen verbatim. The string is the one LocationSensor
        // sends for the same switch, so a screen that draws both sensors reads
        // one vocabulary.
        //
        // No authorization is consulted, unlike LocationSensor, because there
        // is none to consult: the compass asks for nothing, so a `.denied` it
        // receives cannot be about this app's permission. See `armed`.
        if (error as? CLError)?.code == .denied {
            armed = true
            send(["available": false, "error": "location services are off"])
            return
        }
        send(["available": false, "error": error.localizedDescription])
    }

    private func send(_ payload: [String: Any]) {
        guard let report else {
            NSLog("GrMob: heading reading with no host-event channel attached")
            return
        }
        guard let data = try? JSONSerialization.data(withJSONObject: payload),
              let json = String(data: data, encoding: .utf8)
        else { return }
        report("heading", json)
    }
}

File: android/app/src/main/java/com/grmob/app/HeadingSensor.kt
package com.grmob.app

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.util.Log
import org.json.JSONObject
import kotlin.math.abs
import kotlin.math.roundToInt

/**
 * The Android half of grmob's compass (core/heading.go).
 *
 *   core.StartHeading ──▶ "sensor" {kind:"heading", command:"start"} ──▶ here
 *   core.CurrentHeading ◀── "heading" host event ◀── here
 *
 * # Which sensor
 *
 * TYPE_ROTATION_VECTOR, not TYPE_MAGNETIC_FIELD or the long-deprecated
 * TYPE_ORIENTATION. The rotation vector is a *fused* sensor: the platform
 * combines the magnetometer with the gyroscope and accelerometer, which is
 * what makes the reading stable while the phone is moving and what keeps it
 * meaningful when the phone is not flat. Reading the magnetometer directly
 * would mean reimplementing that fusion here, badly.
 *
 * TYPE_GEOMAGNETIC_ROTATION_VECTOR is the lower-power alternative (no
 * gyroscope) and is the fallback when the fused one is absent, which it is on
 * some cheap hardware.
 *
 * # No permission
 *
 * Unlike location, the compass needs none: rotation vector is not a
 * runtime-permission sensor on any API level. That is the whole reason Tier C
 * is a smaller job than the location work that reuses its shape.
 */
object HeadingSensor : SensorEventListener {
    private const val TAG = "GrMobHeading"

    /**
     * Throttle floor. The rotation vector delivers at the rate the sensor
     * hardware runs — SENSOR_DELAY_UI is a hint, not a contract, and on some
     * devices means 50 Hz or more. Every event that reaches Go costs a full
     * render pass, so the stream is cut to roughly 15 Hz here, at the host
     * edge, rather than after crossing the bridge.
     */
    private const val MIN_INTERVAL_MS = 66L

    /**
     * Low-pass smoothing factor for the reported bearing, in [0, 1]: the
     * fraction of each new reading that is taken. A magnetometer jitters by a
     * degree or two even on a still desk, and an unsmoothed needle visibly
     * twitches.
     *
     * 0.25 settles a real turn within a few frames while damping the noise.
     * The filter runs on the *unwrapped* delta (see [smooth]) — averaging 359
     * and 1 arithmetically would give 180, sending the needle due south every
     * time the user crossed north.
     */
    private const val SMOOTHING = 0.25f

    private var manager: SensorManager? = null
    private var sensor: Sensor? = null
    private var report: ((String, String) -> Unit)? = null

    private var running = false
    private var lastSentAt = 0L
    private var smoothed = Float.NaN
    private var accuracyDegrees = -1.0

    private val rotation = FloatArray(9)
    private val orientation = FloatArray(3)

    /**
     * Remembers how to reach this device's sensors and how to answer Go.
     * Call once at startup, before [handle] can be reached by an event.
     *
     * @param report the host→app channel, normally GrMobRuntime::hostEvent.
     */
    fun attach(context: Context, report: (String, String) -> Unit) {
        this.report = report
        manager = context.applicationContext
            .getSystemService(Context.SENSOR_SERVICE) as? SensorManager
        sensor = manager?.let {
            it.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR)
                ?: it.getDefaultSensor(Sensor.TYPE_GEOMAGNETIC_ROTATION_VECTOR)
        }
    }

    /** Dispatches one "sensor" system event. Unknown kinds and commands are dropped. */
    fun handle(data: JSONObject) {
        if (data.optString("kind") != "heading") return
        when (data.optString("command")) {
            "start" -> start()
            "stop" -> stop()
        }
    }

    private fun start() {
        if (running) return
        val m = manager
        val s = sensor
        if (m == null || s == null) {
            // The honest answer for a device with no compass — a tablet, an
            // emulator without sensor emulation. One event, and Go stops
            // waiting for a reading that is never coming.
            send(JSONObject().put("available", false).put("error", "no compass on this device"))
            return
        }
        running = true
        smoothed = Float.NaN
        lastSentAt = 0L
        // SENSOR_DELAY_UI rather than _FASTEST or _GAME: this drives a dial a
        // person is looking at, and the difference between the rates is
        // battery spent on events the throttle above then discards.
        m.registerListener(this, s, SensorManager.SENSOR_DELAY_UI)
    }

    private fun stop() {
        if (!running) return
        running = false
        manager?.unregisterListener(this)
    }

    override fun onSensorChanged(event: SensorEvent) {
        if (!running) return

        SensorManager.getRotationMatrixFromVector(rotation, event.values)
        SensorManager.getOrientation(rotation, orientation)
        // getOrientation answers azimuth in radians over (-pi, pi], measured
        // counter-clockwise-negative from north. Degrees and the fold into
        // [0, 360) both happen on the Go side (core.ReceiveHeading), which is
        // the one place all four hosts funnel through — but the smoothing
        // below needs a continuous number, so the conversion happens here too.
        val degrees = Math.toDegrees(orientation[0].toDouble()).toFloat()
        val next = smooth(degrees)

        val now = System.currentTimeMillis()
        if (now - lastSentAt < MIN_INTERVAL_MS) return
        lastSentAt = now

        val payload = JSONObject()
            .put("magnetic", next.toDouble())
            .put("ts", now)
        if (accuracyDegrees >= 0) payload.put("accuracy", accuracyDegrees)
        send(payload)
    }

    /**
     * Exponential smoothing that respects the seam at north.
     *
     * The filter is applied to the *shortest signed turn* from the last value
     * to the new one rather than to the values themselves, so a step from 359
     * to 1 is treated as the +2 degrees it is instead of as -358. Without
     * this the needle swings the long way round the rose every time the user
     * passes north, which is the single most visible bug a compass can have.
     */
    private fun smooth(degrees: Float): Float {
        if (smoothed.isNaN()) {
            smoothed = degrees
            return smoothed
        }
        var delta = (degrees - smoothed) % 360f
        if (delta > 180f) delta -= 360f
        if (delta <= -180f) delta += 360f
        smoothed += SMOOTHING * delta
        // Keep the accumulator bounded; Go normalises what it is sent, and an
        // unbounded float would lose precision after enough laps.
        if (abs(smoothed) > 720f) smoothed %= 360f
        return smoothed
    }

    override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {
        // Android reports a bucket, not an angle. The mapping to degrees is
        // the platform's own documented guidance for what each bucket means
        // for a compass, so Go and the other two hosts can compare like with
        // like — iOS and Safari both answer in degrees.
        accuracyDegrees = when (accuracy) {
            SensorManager.SENSOR_STATUS_ACCURACY_HIGH -> 5.0
            SensorManager.SENSOR_STATUS_ACCURACY_MEDIUM -> 15.0
            SensorManager.SENSOR_STATUS_ACCURACY_LOW -> 30.0
            // UNRELIABLE means the magnetometer needs the figure-eight
            // calibration wave. Reported as a large angle rather than as
            // "unknown": the app should be able to tell its user the reading
            // is bad, and -1 would say only that nobody measured.
            SensorManager.SENSOR_STATUS_UNRELIABLE -> 180.0
            else -> -1.0
        }
    }

    private fun send(payload: JSONObject) {
        val r = report
        if (r == null) {
            Log.w(TAG, "heading reading with no host-event channel attached")
            return
        }
        r("heading", payload.toString())
    }

    /** Rounded degrees, for logging and tests. */
    internal fun debugBearing(): Int =
        if (smoothed.isNaN()) -1 else ((smoothed % 360f + 360f) % 360f).roundToInt()
}
